save-sandbox/src/main/kotlin/com/saveourtool/save/sandbox/security/CustomAuthenticationBasicConverter.kt

```diff
@@ -0,0 +1,35 @@
+package com.saveourtool.save.sandbox.security
+
+import com.saveourtool.save.utils.AuthenticationDetails
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
+import org.springframework.security.core.Authentication
+import org.springframework.security.web.server.authentication.ServerAuthenticationConverter
+import org.springframework.stereotype.Component
+import org.springframework.web.server.ServerWebExchange
+import reactor.core.publisher.Mono
+
+/**
+ * Implementation of [ServerAuthenticationConverter] that embeds user identity source into [UsernamePasswordAuthenticationToken]
+ */
+@Component
+class CustomAuthenticationBasicConverter : org.springframework.security.web.server.authentication.ServerHttpBasicAuthenticationConverter(),
+ServerAuthenticationConverter {
+    /**
+     * Convert exchange, received from gateway into UsernamePasswordAuthenticationToken, specify source identity, laid
+     * by gateway into X-Authorization-Source header
+     */
+    @Suppress("TOO_MANY_LINES_IN_LAMBDA")
+    override fun convert(exchange: ServerWebExchange): Mono<Authentication> = super.convert(exchange).map { authentication ->
+        val name = (authentication as UsernamePasswordAuthenticationToken).principal as String
+        val source = exchange.request.headers["X-Authorization-Source"]?.firstOrNull()
+        UsernamePasswordAuthenticationToken(
+            "$source:$name",
+            authentication.credentials as String
+        ).apply {
+            details = AuthenticationDetails(
+                id = -1,
```

Comment:
To avoid confusion with integer literals (and sometimes avoid integer overflow), it's better to use the `L` suffix with all `Long` literals, such as `-1L`.

Reply:
Changed